### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by HanOostdijk** [statistiekcbs/cbsccb#1]
In the LinkedIn announcement of this package you asked for feedback. So here it is ;)

Nice and clean interface.
Offering the query possibility in this way was new for me but works quite well.
I found no errors while testing.

Remarks/questions:

- `cbs4_get_toc` is identical to `cbs4_get_datasets`. 
Then define it as such:  `cbs4_get_toc` <- `cbs4_get_datasets` (?)
- after ` meta <- cbs4_get_metadata(id="60006") ` the expression `print(meta)` suggests
` 'For more info use meta$.' ` . This is a little confusing when you don't know what to expect. 
I would advise `str(meta)` or add  `names(meta)` . Nice function by the way!
- the help for  ` cbs4_get_metadata ` uses the term `SN table`. Only after some time 'viel het kwartje'.
- the function  ` cbs4_get_metadata ` shows that a cache is used; 
this could also be done for `cbs4_get_datasets`
- with all data retrieval functions `cbs4_download` , `cbs4_get_data` and `cs4_get_observations` the data is first saved on disk. For a monthly report this does not seem to be necessary: when a program is made to extract and report the data these disk data is no longer needed. But yes: also in these cases they can be necessary for an audit trail.
- related to the query possibilities:
    - from the `$filter` functions only `eq` and `contains` are supported now?
    - I do not understand the `eq` function.  I understand that `a = 34 ` can be written as ` equal(a, 34) ` but this function has three possible arguments. For me an example that uses all arguments would be helpful.
    - the function `cbsodata4:::get_query` has an argument `select` but I could not find that is used by the three data retrieval functions. This is no big deal because the `Observations` table is always 'small' (i.e. 'long').

**Comment by edwindj** [comment]
Thanks for your elaborate feedback, much appreciated!

- `cbs4_get_toc`: good remark, and it is exactly defined as such: <https://github.com/statistiekcbs/cbsodata4/blob/master/R/cbs4_get_datasets.R>. The documentation however treats it as a different function. I supply this function for users who are familiar with `cbsodataR`.

- `meta`, will adopt your suggestion

- `SN`, will adopt your suggestion, (because young users have more trouble finding "kwartjes" for documentation enlightenment :-) )

- `cache` for `cbs4_get_datasets()`. I think that it is a good suggestion, though in practice it won't be used that much. Maybe for very long running scripts (> 1day)  this may be relevant.

- You are correct that it is not necessary for most usages and my first implementations (in `cbsodataR`) did not store the result of the retrieval. This introduced the problem that in some cases the selection is too big to retrieve in one webapi call and that in some cases the result was too big for memory. The programming logic quickly became complex. So I implement data retrieval in the same manner for all scenarios, to make it more stable. The only problem now it that a user could select a table that too big for memory. I think an extra argument that checks on that could be nice. The user would in that case get a warning/message that all data has been downloaded in the directory, but not loaded into memory (or something like that). Any other suggestions are appreciated :-)

- query:

    - yes, only `eq` and `contains` are supported now. This is for the following reasons: these are the main ways of restricting/selecting data downloads and these kind of queries are also generated by the CBS odata web app where a users manually selects categories and measures.  It would be very easy to add `startsWith` and `endsWidth` (implementation is almost done). Other odata query possibility I consider advanced usage, and it would take quite some effort to support this in the same manner as eq and contains. Maybe a simple fix would be that a user could supply and extra filter argument/text that will be added/pasted to the generated query string. 

   -  I understand. I find one of the most challenging (and rewarding) things of building packages like these to provide an 'interface' that works as natural as possible, and still is powerful, so a lot of time goes into crafting and rewriting. `contains` and `eq` are an example of that. They were created to make it easy to make selections in dimensions. So

```r
 cbs4_get_data("84287NED", Perioden = c("2019MM12", "2020MM01"))
 # exact same query, but a bit more verbose
 # cbs4_get_data("84287NED", Perioden = eq(c("2019MM12", "2020MM01")))
```
changes internally `Perioden = c("2019MM12", "2020MM01")` into
` eq("2019MM12", "2020MM01","Perioden")` which in the odata query is expanded into
`Perioden eq '2019MM12' or Perioden eq '2020MM01`. This works identical for `contains`. This
has the weird/nasty side effect that for a programmer who knows odata syntax, the order of the arguments is reversed: in odata a filter query on months would be "contains('Perioden', 'MM') and in cbsodata4 it is 'contains('MM', 'Perioden'). That is unfortunate but I think the syntax in R is now more readable since you will specify the query as `Perioden = contains('MM')`.  The third argument is an implementation detail that I would like to hide, but it internally checks if you query is empty or not. So if you try to filter on an non-existing category it will warn you.

   - good question: this is a relic of `cbsodataR` in which it `$select` was necessary to make a selection on measures. Doubting if it should be supported (it is valid odata syntax) or not. Currently not tested (so buggy).

**Comment by HanOostdijk** [comment]
I like your idea 

> Maybe a simple fix would be that a user could supply and extra filter argument/text that will be added/pasted to the generated query string. 

That extra argument could even be a full OData4 query with `$filter`, `$select` , `...` !
The only problem would be how to combine the two methods of specification (the 'current' one and the OData4 way) when both are used. My preference would be: if you specify it in the new method then any 'current' specification is ignored. 

In that way you can use the full OData4 query language (as far as that is supported) in combination with the infrastructure that you set up.

I will leave the issue open for now, but I think my points are very well addressed.
If you would rather have the issue closed, then please do so.
